tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/aux/ProtoSeqOps.scala

```diff
@@ -0,0 +1,268 @@
+        (ns, cell :: createdCells)
+      }
+
+    // attach the cells to the proto sequence, do not track this in SMT
+    val nextState = afterMk.updateArena(_.appendHasNoSmt(protoSeq, cells: _*))
+    nextState.setRex(protoSeq.toNameEx)
+  }
+
+  /**
+   * Get the capacity of the proto sequence, that is, the number of cells allocated for the sequence elements.
+   *
+   * @param arena
+   *   current arena
+   * @param protoSeq
+   *   a proto sequence
+   * @return
+   *   the number of cells that can fit into the sequence
+   */
+  def capacity(arena: Arena, protoSeq: ArenaCell): Int = {
+    arena.getHas(protoSeq).length
+  }
+
+  /**
+   * Retrieve the elements attached to the proto sequence. This is always a sequence of the [[capacity]]. If you need to
+   * access a single element, use the method [[at]].
+   *
+   * @param arena
+   *   the current arena
+   * @param protoSeq
+   *   a proto sequence
+   * @return
+   *   the sequence of cells that are attached to the proto sequence
+   */
+  def elems(arena: Arena, protoSeq: ArenaCell): Seq[ArenaCell] = {
+    arena.getHas(protoSeq)
+  }
+
+  /**
+   * Retrieve the ith element of the proto sequence. This operation does not produce any SMT constraints and it should
+   * be used as much as possible instead of [[get]].
+   *
+   * @param arena
+   *   the current arena
+   * @param indexBase1
+   *   a fixed index within the capacity
+   * @return
+   *   the cell that is attached to the index
+   */
+  def at(arena: Arena, protoSeq: ArenaCell, indexBase1: Int): ArenaCell = {
+    val protoElems = elems(arena, protoSeq)
+    if (indexBase1 < 1 || indexBase1 > protoElems.length) {
+      throw new IllegalArgumentException(
+          s"Implementation bug: Accessing proto sequence of size ${protoElems.length} with index ${indexBase1}")
+    } else {
+      protoElems(indexBase1 - 1)
+    }
+  }
+
+  /**
+   * Rewrite the index expression and access the element of the proto sequence. This operation produces `O(capacity)`
+   * SMT constraints and it should be avoided, when possible. If the index is an integer literal, use [[at]].
+   *
+   * @param picker
+   *   an instance of CherryPick that is used to pick the values
+   * @param state
+   *   a symbolic state to start with
+   * @param defaultValue
+   *   the value to return if the proto sequence has capacity of zero
+   * @param indexBase1Ex
+   *   an index expression (indices are starting with 1)
+   * @return
+   *   the new symbolic state that contains the retrieved cell as the expression
+   */
+  def get(picker: CherryPick, defaultValue: ArenaCell, state: SymbState, protoSeq: ArenaCell,
+      indexBase1Ex: TlaEx): SymbState = {
+    val protoElems = elems(state.arena, protoSeq)
+    val capacity = protoElems.size
+    if (capacity == 0) {
+      // The sequence is statically empty. Return the default value.
+      // Most likely, this expression will be never used as a result.
+      state.setRex(defaultValue.toNameEx)
```

Comment:
Does it risk inconsistency that an out of bounds access on an empty sequence will produce the constant default value, but out of bounds access on any other sequence would leave the result unrestricted?

Reply:
The users should not rely on the implementation details. When a sequence (or a function) is accessed outside of its domain, the only thing that we know is that the resulting value can be arbitrary. It does not have to be arbitrary though. It is a small optimization here, as the solver should not worry about the value.